20220919 properties (#37)
NLnetLabs/draft-toorop-dnsop-dns-catalog-zones: draft-ietf-dnsop-dns-catalog-zones.md

However, at least one is still required so that catalog zones are syntactically correct DNS zones.
A single NS RR with a NSDNAME field containing the absolute name "invalid." is RECOMMENDED [@!RFC2606].

## Global Properties

Apart from catalog zone metadata stored at the apex (NS, SOA and the like), catalog zone information is stored in the form of "properties".

This specification defines a number of so-called properties that implementations MUST support,
as well as a mechanism to allow implementors to store additional information in the catalog zone (= extension properties, (#customproperties)).
The meaning of such extension properties is determined by the implementation in question.

> Properties can occur both on the global level, or be scoped to apply only to a specific member zone.

Reviewer: ```suggestion
Some properties are defined at the global level; others are scoped to apply only to a specific member zone.
```

for authoritative nameserver management only and are not intended for general
querying via recursive resolvers.

## Global Properties

> Apart from catalog zone metadata stored at the apex (NS, SOA and the like), catalog zone information is stored in the form of "properties".

Reviewer 2: ```suggestion
Apart from catalog zone metadata stored at the apex (NS, SOA and the like), catalog zone information is stored in the form of "properties".
Catalog consumers SHOULD ignore properties it does not understand.
```

Reviewer: it -> they

> This specification defines a number of so-called properties that implementations MUST support,
> as well as a mechanism to allow implementors to store additional information in the catalog zone (= extension properties, (#customproperties)).

Reviewer 2: ```suggestion

This specification defines a few optional properties (in (#cooproperty) and (#groupproperty)) and one obligatory property (in (#versionproperty)),
as well as a mechanism to allow implementors to store additional information in the catalog zone (= extension properties, (#customproperties)).
```

The meaning of such extension properties is determined by the implementation in question.

Some properties are defined at the global level; others are scoped to apply only to a specific member zone.

> This section deals with global properties; member-specific properties are described in (#properties).
> 
> More properties may be defined in future documents.
> 
> ### Schema Version (`version` property)

Reviewer 2: ```suggestion
This document defines a single mandatory global property in (#version). Member-specific properties are described in (#properties).

More properties may be defined in future documents.

### Schema Version (`version` property) {#version}
```

The catalog zone schema version is specified by an integer value embedded in a TXT RR named `version.$CATZ`.

> All catalog zones MUST have a TXT RRset named `version.$CATZ` with at least one RR.

Reviewer 3: There can be only one schema version in a catalog zone. Multiple version RRs will make the catalog invalid.

Reviewer: ```suggestion
All catalog zones MUST have a TXT RRset named `version.$CATZ` with exactly one RR.
```

> Catalog consumers MUST NOT apply catalog zone processing to zones without the expected value in one of the RRs in the `version.$CATZ` TXT RRset, but they may be transferred as ordinary zones.
> For this memo, the value of one of the RRs in the `version.CATZ` TXT RRset MUST be set to "2", i.e.

Reviewer 2: ```suggestion
Catalog consumers MUST NOT apply catalog zone processing to zones without the expected value in the `version.$CATZ` TXT RR, but they may be transferred as ordinary zones.
For this memo, the value of the `version.CATZ` TXT RR MUST be set to "2", i.e.:
```

``` dns-zone
version.$CATZ 0 IN TXT "2"

> ```

Reviewer 3: A global ext property is an incompatible change. I think we should consider bumping the version since processing vs ignoring the global ext property might lead to unexpected results.

Reviewer 2: I don't this is necessary.

In the Description section it states: "Catalog consumers SHOULD ignore any RR in the catalog zone which is meaningless or useless to the implementation.", so current version 2 implementation should already ignore global `.ext` properties.

Furthermore, in Section Custom Properties (*.ext properties): "The meaning of the custom properties described in this section is determined by the implementation alone, without expectation of interoperability."

Also, we haven't had any global property yet (except for the `version` property).

@mind04 @Habbie @libor-peltan-cznic @peterthomassen you agree?

Reviewer 4: I think we should stick with version "2".

Reviewer: "2" works for me.

### Schema Version (`version` property)

The catalog zone schema version is specified by an integer value embedded in a TXT RR named `version.$CATZ`.

> All catalog zones MUST have a TXT RRset named `version.$CATZ` with exactly one RR.

Reviewer 2: The "at least one RR" was to allow for backwards compatibility. For example version 3 introduces some new properties that can be safely ignored by consumers of catalogs of version 2. The producer can in that case put in two RRs in the set, one for 2 and one for 3.

Reviewer 3: > The "at least one RR" was to allow for backwards compatibility. For example version 3 introduces some new properties that can be safely ignored by consumers of catalogs of version 2. The producer can in that case put in two RRs in the set, one for 2 and one for 3.

When the consumer can safely ignore the new property, there is no need to bump the the version number. Version changes are only necessary for non backwards compatible changes. Because of this you never require more than one version record in a zone.

Reviewer 4: I agree that just adding new properties does not require bumping new version number. However, what if some (now unforeseen) changes to catalog zone structure will eventually introduce incompatible change? In that case, we would need to bump the version string number to "3" or more. And it that situation, it might still be useful to generate catalogs that can be interpreted both as "3" and "2". I support keeping the possiblity of multiple version strings, that means the "at least" phrase in this case.

Reviewer: While I can't prove a negative, I find it hard to imagine a backwards incompatible change where we could still generate a catalog that works for both versions. I'm open to leaving room for "both" but I still prefer "one version".

Reviewer 2: I find it hard to imagine that too, but to be on the safe side it would be better to maintain "at least" as it also allows to specify a version 3 that is not compatible with version 2 (by having a single version RR). But not the other way around. An "exactly one" does not allow for catalog producers to generate catalogs suitable for version 2 and 3.

Reviewer: Thought about it some more - such a mixed-version catalog would necessarily be so complex that nobody should be writing the code for it, producer or consumer. So, let's please stick to one version number in a catalog. This makes life simpler for everybody.

Reviewer 2: Are you sure? A consumer just supports 1 version and looks for it in the RRset.
A producer can either produce for multiple version, or for a single version.

Reviewer 4: (Even the consumer might be able to consume multiple versions of catalog if needed.)

What if the incompatible change, which will eventually require bumping the version number, would only relate to some (marginal) catalog features? In that case, simple catalog zones might be version-universal and deserving multi-version.

I don't see whom would restriction to single version string would make "life simpler". Catalog generator can add version fields to its liking (just at least one), interpret can simply search for any known version listed. This is not difficult to implement for anyone.

Reviewer: > What if the incompatible change, which will eventually require bumping the version number, would only relate to some (marginal) catalog features? In that case, simple catalog zones might be version-universal and deserving multi-version.

Again, I have trouble imagining a situation fitting this description. If it's incompatible, you can't leave "2" in there. If it's marginal, it doesn't need "3".

However, we are trying to predict the future here. My prediction is that a single number will be simpler in the future. Your prediction is that supporting a range will be simpler in the future. We don't know who will be right :)

Reviewer 2: Anyway, both options are possible in the future anyway. We can always in an standard defining a new "backwards compatible" catalog zone specify that it's version is specified in a new property `newversion.$CATZ TXT "1"` ;)

Reviewer: That is an excellent point, that I'd like to hijack to argue for a single version RR today.

Reviewer 4: I surrender and accept also forcing a single version string. In that case, the rest of the paragrph needs re-phrasing, too.

Reviewer 2: Ok won't change the already committed change, but alter the rest of the paragraph

Custom properties can occur both globally, or for a specific member zone.
To prevent a name clash with future properties, such properties should be represented below the label `ext`.

`ext` is not a placeholder, so a custom property would have domains names as follows:

```
<your-property>.ext.$CATZ                   # for a global custom property
<your-property>.ext.<unique-N>.zones.$CATZ  # for a member zone custom property
```

`<your-property>` may consist of one or more labels.

Implementations MAY use such properties on the member zone level to store additional information about member zones,
for example to flag them for specific treatment (such as ...).

Further, implementations MAY use custom properties on the global level to store additional information about the catalog zone itself.
While there may be many use cases for this, a plausible one is to store default values for custom properties on the global level,
then overriding them using a property of the same name on the member level (= under the `ext` label of the member node) if so desired.
A property description should clearly say what semantics apply, and whether a property is global, member, or both.

> [NOTE by PTH: This provides no way of *removing* a global default setting via an override at the member zone level.
> While this is not this specification's ceorn, I was wondering if this has been considered?]
> 

Reviewer 2: ```suggestion
```

I think we have by this time all considered it. Inline with the overriding a property text above, an implementation can specify how a property is "removed"/"cleared" at the member zone level.

Reviewer: this NOTE can go!

> The meaning of the custom properties described in this section is determined by the implementation alone, without expectation of interoperability.

Reviewer 2: ```suggestion
The meaning of the custom properties described in this section is determined by the implementation alone, without expectation of interoperability.
A catalog consumers SHOULD ignore custom properties it does not understand.
```

Reviewer: consumers -> consumer

Reviewer 2: Thanks! Fixed that!

serial.<unique-3>.zones.$CATZ 0 IN SERIAL 2020112405
```

## Custom Properties (`*.ext` properties) {#customproperties}

Implementations and operators of catalog zones may choose to provide their own properties.
Custom properties can occur both globally, or for a specific member zone.
To prevent a name clash with future properties, such properties should be represented below the label `ext`.

`ext` is not a placeholder, so a custom property would have domains names as follows:

```
<your-property>.ext.$CATZ                   # for a global custom property
<your-property>.ext.<unique-N>.zones.$CATZ  # for a member zone custom property
```

`<your-property>` may consist of one or more labels.

Implementations MAY use such properties on the member zone level to store additional information about member zones,
for example to flag them for specific treatment (such as ...).

Further, implementations MAY use custom properties on the global level to store additional information about the catalog zone itself.
While there may be many use cases for this, a plausible one is to store default values for custom properties on the global level,

> then overriding them using a property of the same name on the member level (= under the `ext` label of the member node) if so desired.

Reviewer: ```suggestion
then overriding them using a property of the same name on the member level (= under the `ext` label of the member node) if so desired.
A property description should clearly say what semantics apply, and whether a property is global, member, or both.
```

serial.<unique-3>.zones.$CATZ 0 IN SERIAL 2020112405
```

## Custom Properties (`*.ext` properties) {#customproperties}

Implementations and operators of catalog zones may choose to provide their own properties.
Custom properties can occur both globally, or for a specific member zone.
To prevent a name clash with future properties, such properties should be represented below the label `ext`.

`ext` is not a placeholder, so a custom property would have domains names as follows:

```
<your-property>.ext.$CATZ                   # for a global custom property

> <your-property>.ext.<unique-N>.zones.$CATZ  # for a member zone custom property

Reviewer 3: Suggestion:
```ext.<unique-N>.zones.$CATZ                                   # will clear al defined global custom properties for <unique-N>```

Reviewer: What content does one put in `ext....` to clear?

Reviewer 2: I would prefer to leave that out in this draft to be honest. `.ext` options are implementation specific anyway. If an implementation defines an option it can also define how to clear it.

Reviewer: I agree, and pt01, at my request, did an update last week to make clear that global properties are not member-default properties unless their definition says they are.

Reviewer 2: Is there also a pt02?

Reviewer: Sorry, I know @peterthomassen as pt01 :)
